projects/bms_carrier/src/current_sense.c

```diff
@@ -1,6 +1,82 @@
 #include "current_sense.h"
 
-// TODO(SOFT-172): Actually implement this method
+#include "ads1259_adc.h"
+#include "bms.h"
+#include "exported_enums.h"
+#include "fault_bps.h"
+#include "log.h"
+#include "soft_timer.h"
+
+static Ads1259Storage s_storage;
+static bool s_is_charging;
+
+static void prv_ads_error_cb(Ads1259StatusCode code, void *context) {
+  if (code == ADS1259_STATUS_CODE_OUT_OF_RANGE) {
+    LOG_WARN("ADS1259 ERROR: OUT OF RANGE\n");
+  } else if (code == ADS1259_STATUS_CODE_CHECKSUM_FAULT) {
+    LOG_WARN("ADS1259 ERROR: CHECKSUM FAULT\n");
+  }
+
+  if (code != ADS1259_STATUS_CODE_OK) {
+    fault_bps_set(EE_BPS_STATE_FAULT_CURRENT_SENSE);
+  } else {
+    fault_bps_clear(EE_BPS_STATE_FAULT_CURRENT_SENSE);
+  }
+}
+
+// returns value in centiamps
+static int16_t prv_voltage_to_current(double reading) {
+  // current = voltage * 100, see confluence
+  return (int16_t)(100 * 100 * reading);
+}
+
+static void prv_periodic_ads_read(SoftTimerId id, void *context) {
+  double reading = s_storage.reading;
+  int16_t val = prv_voltage_to_current(reading);
+  CurrentStorage *storage = context;
+  storage->readings_ring[storage->ring_idx] = val;
+  storage->ring_idx = (storage->ring_idx + 1) % NUM_STORED_CURRENT_READINGS;
+  ads1259_get_conversion_data(&s_storage);
+  soft_timer_start_millis(storage->conv_period_ms, prv_periodic_ads_read, context, NULL);
+
+  // update average
+  int32_t sum = 0;
+  for (uint16_t i = 0; i < NUM_STORED_CURRENT_READINGS; i++) {
+    sum += storage->readings_ring[i];
+  }
+  storage->average = sum / NUM_STORED_CURRENT_READINGS;
```

Comment:
Assuming the storage gets zero-initialized at the start, the average calculation will be weighted towards 0 until we've read `NUM_STORED_CURRENT_READINGS` values, so the average won't be accurate for the first ~360ms. This is probably fine, though if we're overcurrent right at the start we won't know for ~360ms (also probably fine, but wanted to make sure you're aware).